parser/parser.y

```diff
@@ -3914,14 +3914,18 @@ CastType:
 		x := types.NewFieldType(mysql.TypeString)
 		x.Flen = $2.(int)
 		x.Charset = charset.CharsetBin
-		x.Collate = charset.CharsetBin
+		x.Collate = charset.CollationBin
 		$$ = x
 	}
 |	"CHAR" OptFieldLen OptBinary OptCharset
 	{
 		x := types.NewFieldType(mysql.TypeString)
 		x.Flen = $2.(int)
 		x.Charset = $4.(string)
+		if x.Charset == "" {
+			x.Charset = charset.CharsetUTF8
```

Comment:
@shenli 
I am not sure whether it's a mistake,
there is no binary option in the[ MySQL5.7 manual](https://dev.mysql.com/doc/refman/5.7/en/cast-functions.html#function_cast).